api/v1alpha1/executionhook_types.go

```diff
@@ -0,0 +1,180 @@
+
+	// If specified, controller only select the containers that are listed from the selected pods based on PodSelector.
+	// Otherwise, all containers of the pods will be selected
+	// +optional
+	ContainerList []string `json:"containerList,omitempty" protobuf:"bytes,2,rep,name=containerList"`
+}
+
+// ExecutionHookStatus defines the observed state of ExecutionHook
+type ExecutionHookStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// This is a list of ContainerExecutionHookStatus, with each status representing
+	// information about how hook is executed in a container, including pod name,
+	// container name, ActionTimestamp, ActionSucceed, etc.
+	// +optional
+	HookStatuses []ContainerExecutionHookStatus `json:"hookStatuses,omitempty" protobuf:"bytes,1,rep,name=hookStatuses"`
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific
+// container in a pod
+type ContainerExecutionHookStatus struct {
+	// This field is required
+	PodName string `json:"podName" protobuf:"bytes,1,opt,name=podName"`
+
+	// This field is required
+	ContainerName string `json:"containerName" protobuf:"bytes,2,opt,name=containerName"`
+
+	// If not set, it is nil, indicating Action has not started
+	// If set, it means Action has started at the specified time
+	// +optional
+	Timestamp *metav1.Time `json:"actionTimestamp,omitempty" protobuf:"bytes,3,opt,name=actionTimestamp"`
+
+	// Succeed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after
+	// ActionTimeoutSeconds passes.
+	// +optional
+	Succeed *bool `json:"actionSucceed,omitempty" protobuf:"varint,4,opt,name=actionSucceed"`
+
+	// The last error encountered when executing the action. The hook controller might
+	// update this field each time it retries the execution.
+	// +optional
+	Error *HookError `json:"error,omitempty" protobuf:"bytes,5,opt,name=error"`
+}
+
+// HookError describes the error occurred from hook execution.
+type HookError struct {
+	// Type of the error
+	// This is required
+	ErrorType ErrorType `json:"errorType" protobuf:"bytes,1,opt,name=errorType"`
+
+	// Error message
+	// +optional
+	Message *string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+
+	// More detailed reason why error happens
+	// +optional
+	Reason *string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+
+	// It indicates when the error occurred
+	// +optional
+	Timestamp *metav1.Time `json:"timestamp,omitempty" protobuf:"bytes,4,opt,name=timestamp"`
+}
+
+// ErrorType defines the type of error occurred from hook execution.
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, AlreadyInProgress, etc.
+const (
+	// The execution hook times out
+	Timeout ErrorType = "Timeout"
```

Comment:
Another reason to consider using the Conditions pattern. This proposal might encourage end users to build a state machine off the error conditions (which as you point out in the comment) are likely to be extended and change. Across the rest of the API we generally avoid this as we've learned that its brittle.

Reply:
Thanks for the suggestion. I would like to revisit this as I iterate on the implementation.

Reply:
@kow3ns If Conditions pattern is used, end user can still add more condition types?